File: app/models/qdm/message_definition.rb
module QDM
  # app/models/qdm/message_definition.rb
  class MessageDefinition < DataElement
    include Mongoid::Document
    embedded_in :patient
    field :url
    field :identifier
    field :version
    field :name
    field :title
    field :replaces
    field :status
    field :experimental
    field :date
    field :publisher
    field :contact
    field :description
    field :useContext
    field :jurisdiction
    field :purpose
    field :copyright
    field :base
    field :parent
    field :event
    field :category
    field :focus
    field :responseRequired
    field :allowedResponse
    field :graph
    field :qdmVersion, type: String, default: '4.0.0'
  end
end

File: app/assets/javascripts/MessageDefinition.js
const mongoose = require('mongoose');
const { DataElementSchema } = require('./basetypes/DataElement');
const Code = require('./basetypes/Code');
const Interval = require('./basetypes/Interval');
const Quantity = require('./basetypes/Quantity');
const DateTime = require('./basetypes/DateTime');
const { ComponentSchema } = require('./Component');
const { FacilityLocationSchema } = require('./FacilityLocation');
const { IdSchema } = require('./Id');
const Any = require('./basetypes/Any');

const [Number, String] = [
  mongoose.Schema.Types.Number,
  mongoose.Schema.Types.String,
];

const MessageDefinitionSchema = DataElementSchema({
  url: Any,
  identifier: Any,
  version: Any,
  name: Any,
  title: Any,
  replaces: Any,
  status: Any,
  experimental: Any,
  date: Any,
  publisher: Any,
  contact: Any,
  description: Any,
  useContext: Any,
  jurisdiction: Any,
  purpose: Any,
  copyright: Any,
  base: Any,
  parent: Any,
  event: Any,
  category: Any,
  focus: Any,
  responseRequired: Any,
  allowedResponse: Any,
  graph: Any,
  qdmVersion: { type: String, default: '4.0.0' },
  _type: { type: String, default: 'MessageDefinition' },

});

module.exports.MessageDefinitionSchema = MessageDefinitionSchema;
module.exports.MessageDefinition = mongoose.model('MessageDefinition', MessageDefinitionSchema);
